Plugins/Flow.Launcher.Plugin.Program/ViewModels/AddProgramSourceViewModel.cs

```diff
@@ -0,0 +1,143 @@
+            set
+            {
+                enabled = value;
+                OnPropertyChanged(nameof(Enabled));
+            }
+        }
+
+        private string location = string.Empty;
+        public string Location
+        {
+            get => location;
+            set
+            {
+                location = value;
+                OnPropertyChanged(nameof(Location));
+            }
+        }
+
+        public ProgramSource Source { get; init; }
+        public IPublicAPI API { get; init; }
+        public string AddBtnText { get; init; }
+        private bool LocationModified = false;
+        private bool StatusModified = false;
+        public bool IsCustomSource { get; init; } = true;
+        public bool IsNotCustomSource => !IsCustomSource;
+
+        public AddProgramSourceViewModel(PluginInitContext context, Settings settings)
+        {
+            API = context.API;
+            Settings = settings;
+            AddBtnText = API.GetTranslation("flowlauncher_plugin_program_add");
+        }
+
+        public AddProgramSourceViewModel(PluginInitContext context, Settings settings, ProgramSource programSource) : this(context, settings)
+        {
+            Source = programSource;
+            enabled = Source.Enabled;
+            location = Source.Location;
+            AddBtnText = API.GetTranslation("flowlauncher_plugin_program_update");
+            IsCustomSource = Settings.ProgramSources.Any(x => x.UniqueIdentifier == Source.UniqueIdentifier);
+
+            this.PropertyChanged += (_, args) =>
+            {
+                switch (args.PropertyName)
+                {
+                    case nameof(Location):
+                        LocationModified = true;
+                        break;
+                    case nameof(Enabled):
+                        StatusModified = true;
+                        break;
+                }
+            };
+        }
+
+        public void Browse()
+        {
+            var dialog = new FolderBrowserDialog();
+            DialogResult result = dialog.ShowDialog();
+            if (result == DialogResult.OK)
+            {
+                Location = dialog.SelectedPath;
+            }
+        }
+
+        public (bool, string) AddProgramSource()
+        {
+            if (!Directory.Exists(Location))
+            {
+                return (false, API.GetTranslation("flowlauncher_plugin_program_invalid_path"));
+            }
+            else if (DuplicateSource(Location))
+            {
+                return (false, API.GetTranslation("flowlauncher_plugin_program_duplicate_program_source"));
+            }
+            else
+            {
+                var source = new ProgramSource(Location, Enabled);
+                Settings.ProgramSources.Insert(0, source);
+                ProgramSetting.ProgramSettingDisplayList.Add(source);
+                return (true, null);
+            }
+        }
+
+        public (bool, string) UpdateProgramSource()
```

Comment:
what do you think about using a Try style?
like `public bool TryUpdateProgramSource(out string newLocation)`

Reply:
Nevermind it seems like that style is used mostly when we expect to throw exception...